File: leet_BTREE_0623_add_row.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        return [
            # self.DFS,
            self.BFS,
        ][0](root, val, depth)

    def BFS(self, root, val, depth) -> Optional[TreeNode]:
        assert root
        if depth == 1:
            node = TreeNode(val)
            node.left = root
            return node
        Q = deque()
        Q.append(root)
        while Q:
            depth -= 1
            if depth == 0:
                return root
            N = len(Q)
            for _ in range(N):
                node = Q.popleft()
                if depth == 1 :
                    L, R = TreeNode(val), TreeNode(val)
                    L.left = node.left
                    R.right = node.right
                    node.left, node.right = L, R
                else:
                    if node.left:
                        Q.append(node.left)
                    if node.right:
                        Q.append(node.right)
        return root

    def DFS(self, root, val, depth) -> Optional[TreeNode]:
        if not root:
            return None
        if depth == 1:
            node = TreeNode(val)
            node.left = root
            return node
        if depth == 2:
            L, R = TreeNode(val), TreeNode(val)
            L.left = root.left
            R.right = root.right
            root.left = L
            root.right = R
            return root
        root.left = self.DFS(root.left, val, depth - 1)
        root.right = self.DFS(root.right, val, depth - 1)
        return root

File: leet_BTREE_0623_add_row.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/// newNode := &TreeNode{Val: 5, Left: nil, Right: nil}

func addOneRow(root *TreeNode, val int, depth int) *TreeNode {
    Solutions := []func (*TreeNode, int, int) *TreeNode {
        // DFS,
        BFS,
    }
    return Solutions[0](root, val, depth)
}

func BFS(root *TreeNode, val int, depth int) *TreeNode {

    // assert root != nil
    if depth == 1 {
        return & TreeNode {Val: val, Left: root, Right: nil}
    }
    Q := []*TreeNode{ root }
    for len(Q) != 0 {
        depth--
        if depth == 0 {
            return root
        }
        var L, R *TreeNode
        N := len(Q)
        for N > 0 {
            node := Q[0]
            Q = Q[1:]
            N--
            if (depth == 1) {
                L = & TreeNode {Val: val, Left:node.Left,   Right:nil}
                R = & TreeNode {Val: val, Left:nil,         Right:node.Right}
                node.Left, node.Right = L, R
            } else {
                if node.Left != nil {
                    Q = append(Q, node.Left)
                }
                if node.Right != nil {
                    Q = append(Q, node.Right)
                }
            }
        }
    }
    return root
}

func DFS(root *TreeNode, val int, depth int) *TreeNode {

    if root == nil {

        return root
    }

    if depth == 1 {

        return & TreeNode {Val: val, Left: root, Right: nil}
    }

    if depth == 2 {
        L := & TreeNode {Val: val, Left:nil, Right:nil}
        R := & TreeNode {Val: val, Left:nil, Right:nil}
        L.Left, R.Right = root.Left, root.Right
        root.Left = L
        root.Right = R
        return root
    }

    root.Left = DFS(root.Left, val, depth - 1)
    root.Right = DFS(root.Right, val, depth - 1)
    return root
}
